src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+            string qop = Auth;
+            if (digestResponse.Parameters.ContainsKey(Qop))
+            {
+                // Check if auth-int present in qop string
+                int index1 = digestResponse.Parameters[Qop].IndexOf(AuthInt);
+                if (index1 != -1)
+                {
+                    // Get index of auth if present in qop string
+                    int index2 = digestResponse.Parameters[Qop].IndexOf(Auth);
+
+                    // If index2 < index1, auth option is available
+                    // If index2 == index1, check if auth option available later in string after auth-int.
+                    if (index2 == index1)
+                    {
+                        index2 = digestResponse.Parameters[Qop].IndexOf(Auth, index1 + AuthInt.Length);
+                        if (index2 == -1)
+                        {
+                            qop = AuthInt;
+                        }
+                    }
+                }
+            }
+
+            // Set cnonce
+            string cnonce = GetRandomAlphaNumericString();
+
+            // Calculate response
+            string a1 = credential.UserName + Colon + realm + Colon + credential.Password;
+            if (digestResponse.Parameters[Algorithm].Contains(Sha256Sess))
+            {
+                a1 = ComputeHash(a1, algorithm) + Colon + digestResponse.Parameters[Nonce] + Colon + cnonce;
+            }
+
+            string a2 = request.Method.Method + Colon + request.RequestUri.PathAndQuery;
+            if (qop == AuthInt)
+            {
+                a2 = a2 + Colon + ComputeHash(request.Content.ReadAsStringAsync().Result, algorithm);
+            }
+
+            string response = ComputeHash(ComputeHash(a1, algorithm) + Colon +
+                                        digestResponse.Parameters[Nonce] + Colon +
+                                        digestResponse.NonceCount.ToString("x8") + Colon +
+                                        cnonce + Colon +
+                                        qop + Colon +
+                                        ComputeHash(a2, algorithm), algorithm);
+
+            // Add response
+            sb.AppendKeyValue(Response, response);
+
+            // Add algorithm
+            sb.AppendKeyValue(Algorithm, algorithm, includeQuotes: false);
+
+            // Add opaque
+            sb.AppendKeyValue(Opaque, digestResponse.Parameters[Opaque]);
+
+            // Add qop
+            sb.AppendKeyValue(Qop, qop, includeQuotes: false);
+
+            // Add nc
+            sb.AppendKeyValue(NC, digestResponse.NonceCount.ToString("x8"), includeQuotes: false);
+
+            // Add cnonce
+            sb.AppendKeyValue(CNonce, cnonce, includeComma: false);
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        private static string GetRandomAlphaNumericString()
+        {
+            using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
```

Comment:
The purpose of the cnonce is to avoid some plaintext attacks, Random will give the same sequence of random numbers for a given seed, so the seed has to be different if the application is restarted to avoid same sequence of numbers. MSDN suggests option to use system time but not sure of the impact on real world scenarios so chose the crypto random.